### ISSUE TITLE
lazy loading samples to memory

### ISSUE THREAD

**Issue by imiskolee** [FluidSynth/fluidsynth#1053]
Hi,team:

We knows fluidsynth supports dynamic samples loader, but I checked the details, I think it's just support load full preset from multi instruments soundfont, but I need lazy load samples when it's used for big soundfont file, it's very important for save application memory.

**Comment by derselbst** [comment]
The setting `synth.dynamic-sample-loading`  is called "dynamic-sample-loading" because it dynamically loads samples when they are needed. It's not clear to me which "details" make you believe differently. The soundfont loader parses all presets and all instrument zones, modulators, etc. And when a noteOn is received, the sampledata is loaded on-the-fly. And when a noteOff is received, this sample will be release again, if it is not used by any other voice or synth instance.

**Comment by mawe42** [comment]
The dynamic-sample-loading doesn't get triggered on noteOn, but when a preset is selected for a channel. It then loads all samples for that preset into memory. As soon as the this preset isn't used by any channel anymore, the samples get dropped from memory again. So imiskolee is correct, it doesn't load or drop individual samples, only whole presets.

**Comment by mawe42** [comment]
@imiskolee Can you explain a little bit more about your use-case? What sort of Soundfont size are we talking about here, and how large are the individual samples? And why are you so memory constrained, and what is your maximum limit?

I think answers to those questions will help us in deciding if it would be worth looking into loading samples individually at noteOn or if there are easier ways to achieve your goals.

**Comment by derselbst** [comment]
> The dynamic-sample-loading doesn't get triggered on noteOn, but when a preset is selected for a channel

Sry, my bad. My memory was fooling me...

**Comment by imiskolee** [comment]
> @imiskolee Can you explain a little bit more about your use-case? What sort of Soundfont size are we talking about here, and how large are the individual samples? And why are you so memory constrained, and what is your maximum limit?
> 
> I think answers to those questions will help us in deciding if it would be worth looking into loading samples individually at noteOn or if there are easier ways to achieve your goals.

single instrument preset ~= 200MB，so  if 16 channels, the total memory need 200 * 16 = 3200MB. 
in fact, just a few notes(and layers) needs on a track(channel) for music ,so lazy load samples can be save memory(like NKI Konkakt).

**Comment by derselbst** [comment]
> single instrument preset ~= 200MB，so if 16 channels, the total memory need 200 * 16 = 3200MB.

Counter questions:
* When targeting Android, why do you have to use soundfonts where a single preset requires 200 MB of sample data?
* And why do you have to assign 16 different presets of this size to each MIDI channel? Given that "just a few notes" are played on a track, you should be able to get away with ProgChanges on let's say 2 or 3 MIDI channels.

**Comment by imiskolee** [comment]
> > single instrument preset ~= 200MB，so if 16 channels, the total memory need 200 * 16 = 3200MB.
> 
> Counter questions:
> 
> * When targeting Android, why do you have to use soundfonts where a single preset requires 200 MB of sample data?
> * And why do you have to assign 16 different presets of this size to each MIDI channel? Given that "just a few notes" are played on a track, you should be able to get away with ProgChanges on let's say 2 or 3 MIDI channels.

yes, I can understand what's you said, but my application is a DAW, 1 channel = 1 track, so ProgChanges can not work.

**Comment by mawe42** [comment]
But Toms first question remains unanswered: why do you need to use a Soundfont with such a large sample size? Wouldn't a smaller Soundfont be much more suited to the platform you are targeting?

I'm not sure how well fully dynamic sample loading would work on Android, especially with all samples loaded on the fly from possibly fairly slow internal or external storage (fairly slow in comparison to modern SSDs). How would you handle the probably very inconsistent latency and jitter that loading from storage will produce? Especially as you plan to build a DAW. My gut feeling is to stay away from dynamic-sample-loading entirely... use a Soundfont with a very small memory footprint that still sounds good and give your users the lowest latency and jitter possible on your Platform.

**Comment by imiskolee** [comment]
> 

the first question: if we are using a high quantity SoundFont , it's will be large. because of multi layers. eg: https://rytmenpinne.wordpress.com/sounds-and-such/salamander-grandpiano/, it's over 200MB on a single instruments.

the dynamic loading latency issue also happened on NKI Kontakt.  but it's not big question on DAW case:
1. when we play a note on midi keyboard, trigger  samples load, the sound will be playing after loaded(high latency),it's not a big problem  when we playing midi keyboard.
3. when we play the project(all midi sequence is static), the first time has low quality also ok, first play equal to trigger application prepare is a very common scene.

I'm not sure I'm expressing myself clearly.

**Comment by derselbst** [comment]
Fluidsynth is a Soundfont2 synth. The soundfont you've linked is SFZ. There is no such thing as "layers" in SF2. Even if you find a SFZ to SF2 converter, I guess you would end up with a single piano preset consisting of multiple instruments. And even if this single preset comprises 200 MB of sample data,  assigning this preset to all 16 channels would only lead to 200 MB memory consumption. And not lead to 3,2 GB as you claimed earlier.


>    when we play a note on midi keyboard, trigger samples load, the sound will be playing after loaded(high latency),it's not a big problem when we playing midi keyboard.

IMO, this would be a huge problem. The strength of fluidsynth is its real-time capability, i.e. you hit a key and you instantly get sound. Given the non-deterministic time  it would take to read the samples from a potentially slow device, decode them and place them into memory, potentially running into an out-of-memory situation causing other memory to be swapped out, would definitely ruin every keyboard live-performance session.

>    when we play the project(all midi sequence is static), the first time has low quality also ok, first play equal to trigger application prepare is a very common scene.

I don't understand this point.

**Comment by mawe42** [comment]
I think he means velocity layers, the Salamander SFZ apparently has 16 different ones. They should be easily mappable to SF2 velocity ranges. But that still doesn't answer the question of *why*. Why do you need to use such a large Soundfont on a memory constrained platform?

> when we play a note on midi keyboard, trigger samples load, the sound will be playing after loaded(high latency),it's not a big problem when we playing midi keyboard.

I concur with Tom... I fail to see how high latency in real-time playback is "not a big problem". As a musician myself, I would consider this a dealbreaker and immediately uninstall your app if it had high and jittery latency.

> when we play the project(all midi sequence is static), the first time has low quality also ok, first play equal to trigger application prepare is a very common scene.

So you would expect the first playback to be jittery and with some timing problems due to the sample-loading, but the second playback would be better because all samples are already in memory? Well... then I don't understand your feature request. You seem to want lazy-loading of samples, but they should stay in memory even after the sample has finished playing? So what happens to your app when one of your users is a skilled player and makes use of a lot of samples in your huge soundfont? Wouldn't you run into memory problems again? And when do you expect the samples to be unloaded again?

To be honest, I'm not sure how fruitful it is to continue this discussion. In my opinion, adding dynamic-sample-loading at noteOn to FluidSynth would be a big and difficult job. It would introduce the possibility of timing problems, jittery and low-quality playback, especially on the platform you are targeting. But if you are willing to take this on, I would definitely have a look at your code and give feedback.

**Comment by imiskolee** [comment]
I have looks another solution `SF3`, looks like `SF3` can be compressed 10x from `SF2`.

**Comment by mawe42** [comment]
SF3 uses Ogg Vorbis compressed samples (a lossy format) to reduce on-disk size of Soundfonts. They get decompressed when loading the samples into memory. So if your goal is to reduce RAM consumption, SF3 does not help you.
